hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/AWSClientConfig.java

```diff
@@ -0,0 +1,231 @@
+import static org.apache.hadoop.fs.s3a.Constants.DEFAULT_SECURE_CONNECTIONS;
+import static org.apache.hadoop.fs.s3a.Constants.DEFAULT_SOCKET_TIMEOUT;
+import static org.apache.hadoop.fs.s3a.Constants.ESTABLISH_TIMEOUT;
+import static org.apache.hadoop.fs.s3a.Constants.MAXIMUM_CONNECTIONS;
+import static org.apache.hadoop.fs.s3a.Constants.MAX_ERROR_RETRIES;
+import static org.apache.hadoop.fs.s3a.Constants.PROXY_DOMAIN;
+import static org.apache.hadoop.fs.s3a.Constants.PROXY_HOST;
+import static org.apache.hadoop.fs.s3a.Constants.PROXY_PASSWORD;
+import static org.apache.hadoop.fs.s3a.Constants.PROXY_PORT;
+import static org.apache.hadoop.fs.s3a.Constants.PROXY_USERNAME;
+import static org.apache.hadoop.fs.s3a.Constants.PROXY_WORKSTATION;
+import static org.apache.hadoop.fs.s3a.Constants.REQUEST_TIMEOUT;
+import static org.apache.hadoop.fs.s3a.Constants.SECURE_CONNECTIONS;
+import static org.apache.hadoop.fs.s3a.Constants.SOCKET_TIMEOUT;
+import static org.apache.hadoop.fs.s3a.Constants.USER_AGENT_PREFIX;
+
+public final class AWSClientConfig {
+  private static final Logger LOG = LoggerFactory.getLogger(AWSClientConfig.class);
+
+  private AWSClientConfig() {
+  }
+
+  public static ClientOverrideConfiguration.Builder createClientConfigBuilder(Configuration conf) {
+    ClientOverrideConfiguration.Builder overrideConfigBuilder =
+        ClientOverrideConfiguration.builder();
+
+    long requestTimeoutMillis = conf.getTimeDuration(REQUEST_TIMEOUT,
+        DEFAULT_REQUEST_TIMEOUT, TimeUnit.SECONDS, TimeUnit.MILLISECONDS);
+
+    if (requestTimeoutMillis > Integer.MAX_VALUE) {
+      LOG.debug("Request timeout is too high({} ms). Setting to {} ms instead",
+          requestTimeoutMillis, Integer.MAX_VALUE);
+      requestTimeoutMillis = Integer.MAX_VALUE;
+    }
+
+    if(requestTimeoutMillis > 0) {
+      overrideConfigBuilder.apiCallAttemptTimeout(Duration.ofMillis(requestTimeoutMillis));
+    }
+
+    initUserAgent(conf, overrideConfigBuilder);
+
+    // TODO: Look at signers. See issue https://github.com/aws/aws-sdk-java-v2/issues/1024
+    //    String signerOverride = conf.getTrimmed(SIGNING_ALGORITHM, "");
+    //    if (!signerOverride.isEmpty()) {
+    //      LOG.debug("Signer override = {}", signerOverride);
+    //      overrideConfigBuilder.putAdvancedOption(SdkAdvancedClientOption.SIGNER)
+    //    }
+
+    return overrideConfigBuilder;
+  }
+
+  /**
+   * Configures the http client.
+   *
+   * @param conf The Hadoop configuration
+   * @return Http client builder
+   */
+  public static ApacheHttpClient.Builder createHttpClientBuilder(Configuration conf) {
+    ApacheHttpClient.Builder httpClientBuilder =
+        ApacheHttpClient.builder();
+
+    httpClientBuilder.maxConnections(S3AUtils.intOption(conf, MAXIMUM_CONNECTIONS,
+        DEFAULT_MAXIMUM_CONNECTIONS, 1));
+
+    httpClientBuilder.connectionTimeout(Duration.ofSeconds(
+        S3AUtils.intOption(conf, ESTABLISH_TIMEOUT, DEFAULT_ESTABLISH_TIMEOUT, 0)));
```

Comment:
Consider moving the intOption call to a temporary variable.

```suggestion
    int connTimeoutMillis = S3AUtils.intOption(conf, ESTABLISH_TIMEOUT, DEFAULT_ESTABLISH_TIMEOUT, 0);
    httpClientBuilder.connectionTimeout(Duration.ofSeconds(connTimeoutMillis));
```

Also applicable on a few other lines, where we have nested method calls in the args.